azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationAction.kt

```diff
@@ -23,5 +36,6 @@ class ScriptGenerationParallelAction(actions: List<List<ScriptGenerationAction>>
     """
 }
 
-fun Action.toScriptGenAction() =
-    this as? ScriptGenerationAction ?: throw IllegalArgumentException("Incompatible action: $this")
+@Suppress("UNCHECKED_CAST")
+fun <E : ScriptGenerationEnvironment> Action.toScriptGenAction() =
+    this as? ScriptGenerationAction<E> ?: throw IllegalArgumentException("Incompatible action: $this")
```

Comment:
⛏️ Usually unchecked casts are a bit of a code smell. Doesn't this work?
```suggestion
fun Action.toScriptGenAction() =
    this as? ScriptGenerationAction<*> ?: throw IllegalArgumentException("Incompatible action: $this")
```

Reply:
That'd return a `ScriptGenerationAction<*>`, surely?

I did consider making `ScriptGenerationEnvironment` have to implement some functions that served only to safely cast `Action` to `ScriptGenerationAction<E>` etc.  But those would need `ScriptGenerationEnvironment` to be type-parameterised on itself (yay, curiously-recurring template pattern!).  I did have similar problems later on that would be solved with a CRTP, so maybe I'll have to do it.

Reply:
> That'd return a ScriptGenerationAction<*>, surely?

Well, the inner type of generics is [erased](https://kotlinlang.org/docs/generics.html#type-erasure) at runtime, so AFAIK there's no need to jump through hoops here unless you actually want to preserve the type of the environment for compile-time checks further downstream.

Reply:
I do :)  Specifically `toScriptGenAction()` returns `ScriptGenerationAction<E>`, and is called by `getBuildActions` which returns `List<ScriptGenerationAction<E>`, and is then fed by `generateWheneverScript` into `generateWheneverScriptFromActions` which takes `List<ScriptGenerationAction<E>>` and calls `getActionScript(environment)` thereby using the knowledge that the `E` of the action lines up with the `E` of the generator.

I could use `filterToInstance` instead of `map(Action::toScriptGenAction)`, I guess?  But that'll silently disappear actions rather than just leading to runtime failures.

Reply:
FYI, I don't think you will ever have _runtime_ failures due to this cast. As I mentioned above, the inner type of the generic is only used at compile time and is erased afterwards (hence the "unchecked cast" warning). At runtime you can only see the outer type, `ScriptGenerationAction<*>`.